src/librustc_mir/util/elaborate_drops.rs

```diff
@@ -535,6 +535,114 @@ impl<'l, 'b, 'tcx, D> DropCtxt<'l, 'b, 'tcx, D>
         })
     }
 
+    /// create a loop that drops an array:
+    ///
+    /// loop-block:
+    ///    can_go = index < len
+    ///    if can_go then drop-block else succ
+    /// drop-block:
+    ///    ptr = &mut LV[len]
+    ///    index = index + 1
+    ///    drop(ptr)
+    fn drop_loop(&mut self,
+                 unwind: Option<BasicBlock>,
+                 succ: BasicBlock,
+                 index: &Lvalue<'tcx>,
+                 length: &Lvalue<'tcx>,
+                 ety: Ty<'tcx>,
+                 is_cleanup: bool)
+                 -> BasicBlock
+    {
+        let use_ = |lv: &Lvalue<'tcx>| Operand::Consume(lv.clone());
+        let tcx = self.tcx();
+
+        let ref_ty = tcx.mk_ref(tcx.types.re_erased, ty::TypeAndMut {
+            ty: ety,
+            mutbl: hir::Mutability::MutMutable
+        });
+        let ptr = &Lvalue::Local(self.new_temp(ref_ty));
+        let can_go = &Lvalue::Local(self.new_temp(tcx.types.bool));
+
+        let one = self.constant_usize(1);
+        let drop_block = self.elaborator.patch().new_block(BasicBlockData {
+            statements: vec![
+                Statement { source_info: self.source_info, kind: StatementKind::Assign(
+                    ptr.clone(), Rvalue::Ref(
+                        tcx.types.re_erased, BorrowKind::Mut,
+                        self.lvalue.clone().index(use_(index))
+                    ),
+                )},
+                Statement { source_info: self.source_info, kind: StatementKind::Assign(
+                    index.clone(), Rvalue::BinaryOp(BinOp::Add, use_(index), one)
+                )},
+            ],
+            is_cleanup,
+            terminator: Some(Terminator {
+                source_info: self.source_info,
+                kind: TerminatorKind::Resume,
```

Comment:
NIT: `Unreachable` here would make it much more obvious this terminator is patched out a bit later.